borsh-derive-internal/src/enum_ser.rs

```diff
@@ -19,7 +23,20 @@ pub fn enum_ser(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
     let mut all_variants_idx_body = TokenStream2::new();
     let mut fields_body = TokenStream2::new();
     let discriminants = discriminant_map(&input.variants);
-    for variant in input.variants.iter() {
+    let has_explicit_discriminants = input
+        .variants
+        .iter()
+        .any(|variant| variant.discriminant.is_some());
+    if has_explicit_discriminants && use_discriminant.is_none() {
+        return Err(syn::Error::new(
+            input.ident.span(),
+            "You have to specify `#[borsh_use_discriminant=true]` or `#[borsh_use_discriminant=false]` for all structs that have enum with explicit discriminant",
+        ));
+    }
+
+    for (variant_idx, variant) in input.variants.iter().enumerate() {
+        let _variant_idx =
+            u8::try_from(variant_idx).expect("up to 256 enum variants are supported");
```

Comment:
Let's assert the `input.variants.len()` once before the loop instead of doing it in the loop.